src/exo/cursors.py

```diff
@@ -0,0 +1,228 @@
+from __future__ import annotations
+
+import weakref
+from dataclasses import dataclass
+from enum import Enum
+from typing import Union, List, Iterable
+from weakref import ReferenceType
+
+from .API_types import ProcedureBase
+from .LoopIR import LoopIR
+
+
+class CursorKind(Enum):
+    Node = 0  # Pointing directly at a node
+    # BlockFront = 1  # Gap before the first statement of a block
+    # BlockEnd = 2  # Gap after the last statement of a block
+    GapBefore = 3  # Gap before the node in a block
+    GapAfter  = 4  # Gap after the node in a block
+
+
+class ExoIR(Enum):
+    Assign = 0
+    Reduce = 1
+    WriteConfig = 2
+    Pass = 3
+    If = 4
+    For = 5
+    Alloc = 6
+    Call = 7
+    WindowStmt = 8
```

Comment:
This seems redundant with `LoopIR` itself in a way that can fall out of sync.

Reply:
i.e. can we just use references to the classes themselves rather than an enumeration?

Reply:
This was patched in for the idea of making Cursor more like a QAST, to allow flexible user introspection, but is currently unused. Currently, `_node()` is a reference to a LoopIR, but we probably don't want users to introspect on LoopIR, so I guess we need something similar to QAST.